File: Leetcode/Python/_637.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        level_map = {}
        output = []

        def findmax(root, level):
            if level not in level_map:
                level_map[level] = [root.val]
            else:
                level_map[level].append(root.val)

            level += 1
            if root.left:
                findmax(root.left, level)
            if root.right:
                findmax(root.right, level)
            if root.left == None and root.right == None:
                return root


        findmax(root,1)   

        for key, values in level_map.items():
            output.append(sum(values) / len(values))
        return output
    
class Solution:
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        q = deque([root])
        result = []
        
        while q:
            n = len(q)
            level_sum = 0
            
            for _ in range(n):
                node = q.popleft()
                level_sum += node.val
                
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            result.append(level_sum / n)
        
        
        return result

File: Leetcode/cpp/637.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        std::vector<double> output;
        std::deque<TreeNode*> dq{root};
        while (!dq.empty()) {
            int size = dq.size();
            double acc = 0.0;
            for (int i = 0; i < size; ++i) {
                acc += dq.front()->val;
                TreeNode* node = dq.front();
                dq.pop_front();
                if (node->left) {
                    dq.push_back(node->left);
                }
                if (node->right) {
                    dq.push_back(node->right);
                }
            }
            output.push_back(acc / (double) size);
        }
        return output;
    }
};
